Bottom: 087fa484b9e8cf1c0373fd0d744e69de6847b0c4
Top:    1e6b91905c88d6e403ffbfb8dea1b044da1b549c
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-25 17:37:04 +0100

draw.c: Split "Gathering" routines from real drawing routines (Lines)


---

diff --git a/src/draw.c b/src/draw.c
index b31f6fd..279108c 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -89,7 +89,6 @@ static const BoxType *clip_box = NULL;
 static void DrawEverything (BoxTypePtr);
 static void DrawPPV (int group, const BoxType *);
 static int DrawLayerGroup (int, const BoxType *);
-static void DrawLineLowLevel (LineTypePtr);
 static void DrawRegularText (LayerTypePtr, TextTypePtr);
 static void DrawPolygonLowLevel (PolygonTypePtr);
 static void DrawArcLowLevel (ArcTypePtr);
@@ -478,6 +477,42 @@ hole_counting_callback (const BoxType * b, void *cl)
   return 1;
 }
 
+static void
+_draw_line (LineType *line)
+{
+  gui->set_line_cap (Output.fgGC, Trace_Cap);
+  if (TEST_FLAG (THINDRAWFLAG, PCB))
+    gui->set_line_width (Output.fgGC, 0);
+  else
+    gui->set_line_width (Output.fgGC, line->Thickness);
+
+  gui->draw_line (Output.fgGC,
+		  line->Point1.X, line->Point1.Y,
+		  line->Point2.X, line->Point2.Y);
+}
+
+static void
+draw_line (LayerType *layer, LineType *line)
+{
+  if (TEST_FLAG (SELECTEDFLAG | FOUNDFLAG, line))
+    {
+      if (TEST_FLAG (SELECTEDFLAG, line))
+        gui->set_color (Output.fgGC, layer->SelectedColor);
+      else
+        gui->set_color (Output.fgGC, PCB->ConnectedColor);
+    }
+  else
+    gui->set_color (Output.fgGC, layer->Color);
+  _draw_line (line);
+}
+
+static int
+line_callback (const BoxType * b, void *cl)
+{
+  draw_line ((LayerType *) cl, (LineType *) b);
+  return 1;
+}
+
 static int
 rat_callback (const BoxType * b, void *cl)
 {
@@ -508,7 +543,7 @@ rat_callback (const BoxType * b, void *cl)
                      w * 2, w * 2, 0, 360);
     }
   else
-    DrawLineLowLevel ((LineType *) rat);
+    _draw_line ((LineType *) rat);
   return 1;
 }
 
@@ -528,7 +563,7 @@ draw_element_package (ElementType *element)
   /* draw lines, arcs, text and pins */
   ELEMENTLINE_LOOP (element);
   {
-    DrawLineLowLevel (line);
+    _draw_line (line);
   }
   END_LOOP;
   ARC_LOOP (element);
@@ -931,13 +966,6 @@ DrawRats (BoxTypePtr drawn_area)
 }
 
 static int
-line_callback (const BoxType * b, void *cl)
-{
-  DrawLine ((LayerTypePtr) cl, (LineTypePtr) b);
-  return 1;
-}
-
-static int
 arc_callback (const BoxType * b, void *cl)
 {
   DrawArc ((LayerTypePtr) cl, (ArcTypePtr) b);
@@ -1088,29 +1116,6 @@ GatherPadName (PadTypePtr Pad)
 }
 
 /* ---------------------------------------------------------------------------
- * lowlevel drawing routine for lines
- */
-static void
-DrawLineLowLevel (LineTypePtr Line)
-{
-  if (Gathering)
-    {
-      AddPart (Line);
-      return;
-    }
-
-  gui->set_line_cap (Output.fgGC, Trace_Cap);
-  if (TEST_FLAG (THINDRAWFLAG, PCB))
-    gui->set_line_width (Output.fgGC, 0);
-  else
-    gui->set_line_width (Output.fgGC, Line->Thickness);
-
-  gui->draw_line (Output.fgGC,
-		  Line->Point1.X, Line->Point1.Y,
-		  Line->Point2.X, Line->Point2.Y);
-}
-
-/* ---------------------------------------------------------------------------
  * lowlevel drawing routine for text objects
  */
 void
@@ -1164,7 +1169,7 @@ DrawTextLowLevel (TextTypePtr Text, int min_line_width)
 	      newline.Point1.Y += Text->Y;
 	      newline.Point2.X += Text->X;
 	      newline.Point2.Y += Text->Y;
-	      DrawLineLowLevel (&newline);
+	      _draw_line (&newline);
 	    }
 
 	  /* move on to next cursor position */
@@ -1320,19 +1325,9 @@ DrawPadName (PadTypePtr Pad)
 void
 DrawLine (LayerTypePtr Layer, LineTypePtr Line)
 {
-  if (!Gathering)
-    {
-      if (TEST_FLAG (SELECTEDFLAG | FOUNDFLAG, Line))
-	{
-	  if (TEST_FLAG (SELECTEDFLAG, Line))
-	    gui->set_color (Output.fgGC, Layer->SelectedColor);
-	  else
-	    gui->set_color (Output.fgGC, PCB->ConnectedColor);
-	}
-      else
-	gui->set_color (Output.fgGC, Layer->Color);
-    }
-  DrawLineLowLevel (Line);
+  assert (Gathering);
+
+  AddPart (Line);
 }
 
 /* ---------------------------------------------------------------------------
@@ -1359,7 +1354,7 @@ DrawRat (RatTypePtr Line)
       AddPart(&b);
     }
   else
-    DrawLineLowLevel ((LineType *) Line);
+    DrawLine (NULL, (LineType *) Line);
 }
 
 /* ---------------------------------------------------------------------------
@@ -1545,7 +1540,7 @@ DrawElementPackage (ElementTypePtr Element)
 
   ELEMENTLINE_LOOP (Element);
   {
-    DrawLineLowLevel (line);
+    DrawLine (NULL, line);
   }
   END_LOOP;
   ARC_LOOP (Element);
@@ -1607,7 +1602,7 @@ EraseRat (RatTypePtr Rat)
 		     w * 2, w * 2, 0, 360);
     }
   else
-    DrawLineLowLevel ((LineTypePtr) Rat);
+    _draw_line ((LineTypePtr) Rat);
 }
 
 
@@ -1676,7 +1671,9 @@ ErasePinName (PinTypePtr Pin)
 void
 EraseLine (LineTypePtr Line)
 {
-  DrawLineLowLevel (Line);
+  assert (Gathering);
+
+  AddPart (Line);
 }
 
 /* ---------------------------------------------------------------------------
@@ -1722,7 +1719,7 @@ EraseElement (ElementTypePtr Element)
 {
   ELEMENTLINE_LOOP (Element);
   {
-    DrawLineLowLevel (line);
+    EraseLine (line);
   }
   END_LOOP;
   ARC_LOOP (Element);
